Add guide on writing a parser (#1346)
Siteimprove/alfa: docs/guides/writing-a-parser.md

# Writing a parser

This guide shows how to implement a simple parser using the [alfa-parser](https://github.com/Siteimprove/alfa/tree/main/packages/alfa-parser) package. The guide is based on the work done in [#1339][PR].

## Background

> The `alfa-parser` package contains functionality to make it easy to implement a parser. This package is used internally by Alfa to parse HTML, CSS etc., but it can be used for any parsing scenario.

Reviewer: ```suggestion
The `alfa-parser` package contains functionality to make it easy to implement a parser ([parsers combinators](https://en.wikipedia.org/wiki/Parser_combinator)). This package is used internally by Alfa to parse HTML, CSS etc., but it can be used for any parsing scenario.
```

For this guide we want to create a parser that parses the content of the HTML attribute `autocomplete`. The content of the attribute must follow a set of rules as defined in the [HTML specification](https://html.spec.whatwg.org/multipage/#autofill-detail-tokens).

## Defining the grammar

We refer to the rules defined in the specification as the grammar or the syntax of the `autocomplete` attribute. It must contain a space separated list of tokens in a specific order, some of which may be optional and some may be required depending on the precense of some preceding token. For example the following is an invalid use of the attribute
```HTML
<input autocomplete="section-blue billing home" />
```
This is invalid because the `home` token must be followed by a required token.

We can express the grammar as follows
```
[section] [addressType] (unmodifiable | [modifier] modifiable) [webauthn] EOF
```

> where a token name sorounded by square brackets `[]` means it is an optional token, tokens seperated by a vertical line `|` means exactly one of the tokens must be present and `EOF` is a special token signaling that no more tokens are allowed.

Reviewer: ```suggestion
where a token name surrounded by square brackets `[]` means it is an optional token, tokens separated by a vertical line `|` means exactly one of the tokens must be present and `EOF` is a special token signalling that no more tokens are allowed.
```

Here the possible values for the tokens are not listed, but as an example `addressType` is defined as follows
```

> addressType = "shipping" | "billing"

Reviewer: ```suggestion
addressType ::= "shipping" | "billing"
```
Keeping it closer to [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) syntax

```
meaning that the token can either have the literal value of `shipping` or `billing`.

## Implementing the parser

> Our strategy is to define individual "atomic" parsers for each of the tokens and then use the functions available in the `alfa-parser` package to combine them into the final parser in such a way that it encodes the grammar defined above. The individual parsers will be very simple since they will only check that a token has one of the valid literal values for that token. If it does, it will return an `Ok` result with the parsed token and the rest of the tokens ready to be parsed by the next parser. If it doesn't, it will return an `Err` result with a message and the parsing will terminate.

Reviewer: I'd rephrase this a bit in the different order:
* What is the `Parser` type (consume an input, here an array of tokens (string); return a `result` which is either an `Ok` with both the remaining input and the actual result, or an `Err` with any relevant information). (more or less the last two sentence)
* Writing parser strategy: atomic parsers + combinators.
* Here, atomic parser = string matching.

Start by adding the following imports
```TS
import { Parser } from "@siteimprove/alfa-parser";
import { Err, Ok } from "@siteimprove/alfa-result";
import { Slice } from "@siteimprove/alfa-slice";

const { either, end, option, right } = Parser;
```

> The types from `alfa-result` will be used by the parser to signal success or failure and the `Slice` type is used by the parser to consume an array of tokens in a memory efficient way. The functions, or combinators, from the `Parser` namespace that we will be using have been stored as `const`s for convenience.

Reviewer: ```suggestion
The types from `alfa-result` will be used by the parser to signal success or failure and the `Slice` type is used by the parser to consume an array of tokens in a memory efficient way. The functions, or combinators, from the `Parser` namespace that we will be used have been stored as `const`s for convenience.
```

We will implement two of the "atomic" parsers and leave out the rest as they are very similar. In the actual implementation in [#1339][PR] this was done in a more general way where a function generates the parser from an array of literal values.

```TS

> const segment: Parser<Slice<string>, string, string> = (input: Slice<string>) => {

Reviewer: ```suggestion
const section: Parser<Slice<string>, string, string> = (input: Slice<string>) => {
```

The implemtation of the `addressType` parser is very similar
```TS
const addressType: Parser<Slice<string>, string, string> = (input: Slice<string>) => {
  const token: Option<string> = input.first(); 
  
  if (token.isSome() && (token.get() === "shipping" || token.get() === "billing")) { 
    return Result.of([input.rest(), token.get()]); 
  }
  
  return Err.of(`Expected \`shipping\` or \`billing\`, but got ${token}`); 
}
```

Now let's imagine we also implemented the other parsers in a similar fashion. We are now ready to combine all the parsers using the parser combinators `option`, `either`, `right` and `end` to get the final parser.

```TS
const parse = right(
  option(section), // The attribute should optionally begin with a section token
  right(

>     option(addressType), Next, optionally an addressType token

Reviewer: ```suggestion
    option(addressType), // Next, optionally an addressType token
```

## Defining the grammar

We refer to the rules defined in the specification as the grammar or the syntax of the `autocomplete` attribute. It must contain a space separated list of tokens in a specific order, some of which may be optional and some may be required depending on the precense of some preceding token. For example the following is an invalid use of the attribute
```HTML
<input autocomplete="section-blue billing home" />
```
This is invalid because the `home` token must be followed by a required token.

We can express the grammar as follows
```
[section] [addressType] (unmodifiable | [modifier] modifiable) [webauthn] EOF
```
where a token name surrounded by square brackets `[]` means it is an optional token, tokens separated by a vertical line `|` means exactly one of the tokens must be present and `EOF` is a special token signalling that no more tokens are allowed.

Here the possible values for the tokens are not listed, but as an example `addressType` is defined as follows
```
addressType ::= "shipping" | "billing"
```
meaning that the token can either have the literal value of `shipping` or `billing`.

## Implementing the parser

> Out aim is to write a parser that consumes an array of tokens (in our case strings) and returns a result of either `Ok`, with the result and the rest of the tokens, or an `Err` result with a message containing relevant information on why the parsing failed.

Author: ```suggestion
Our aim is to write a parser that consumes an array of tokens (in our case strings) and returns a result of either `Ok`, with the result and the rest of the tokens, or an `Err` result with a message containing relevant information on why the parsing failed.
```

```
[section] [addressType] (unmodifiable | [modifier] modifiable) [webauthn] EOF
```
where a token name surrounded by square brackets `[]` means it is an optional token, tokens separated by a vertical line `|` means exactly one of the tokens must be present and `EOF` is a special token signalling that no more tokens are allowed.

Here the possible values for the tokens are not listed, but as an example `addressType` is defined as follows
```
addressType ::= "shipping" | "billing"
```
meaning that the token can either have the literal value of `shipping` or `billing`.

## Implementing the parser

Our aim is to write a parser that consumes an array of tokens (in our case strings) and returns a result of either `Ok`, with the result and the rest of the tokens, or an `Err` result with a message containing relevant information on why the parsing failed.

The strategy is:
- write individual atomic parsers to parse each indivial token using string matching

> - combine the atomic parser using the combinators from `alfa-parser` into the final parser in such a way that it encodes the grammar

Reviewer: ```suggestion
- combine the atomic parsers using the combinators from `alfa-parser` into the final parser in such a way that it encodes the grammar
```
